engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -40,367 +40,223 @@
     }
-    if (a instanceof Atom && b instanceof Atom) {
+
+    @Specialization(replaces = "doConversionCached")
+    final boolean doConversionUncached(
+        VirtualFrame frame,
+        Object self,
+        Object that,
+        @Shared("typeOf") @Cached TypeOfNode typeOfNode,
+        @Shared("convert") @Cached InteropConversionCallNode convertNode,
+        @Shared("invoke") @Cached(allowUncached = true) EqualsSimpleNode equalityNode) {
+      var selfType = findType(typeOfNode, self);
+      var thatType = findType(typeOfNode, that);
+      if (findConversions(selfType, thatType) != null) {
+        var result = doDispatch(frame, self, that, selfType, convertNode, equalityNode);
+        return result;
+      }
       return false;
     }
-    if (warnings.hasWarnings(a) || warnings.hasWarnings(b)) {
-      return true;
-    }
-    if (a instanceof EnsoMultiValue || b instanceof EnsoMultiValue) {
-      return true;
-    }
-    return !isPrimitive(a, interop) && !isPrimitive(b, interop);
-  }
-
-  /**
-   * Return true iff object is a primitive value used in some specializations guard. By primitive
-   * value we mean any value that can be present in Enso, so, for example, not Integer, as that
-   * cannot be present in Enso. All the primitive types should be handled in their corresponding
-   * specializations. See {@link
-   * org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode}.
-   */
-  static boolean isPrimitive(Object object, InteropLibrary interop) {
-    return isPrimitiveValue(object)
-        || object instanceof EnsoBigInteger
-        || object instanceof Text
-        || interop.isString(object)
-        || interop.isNumber(object)
-        || interop.isBoolean(object);
-  }
 
-  static boolean isPrimitiveValue(Object object) {
-    return object instanceof Boolean || object instanceof Long || object instanceof Double;
-  }
-
-  static boolean isBigInteger(InteropLibrary iop, Object v) {
-    if (v instanceof EnsoBigInteger) {
-      return true;
-    } else {
-      return !iop.fitsInDouble(v) && !iop.fitsInLong(v) && iop.fitsInBigInteger(v);
+    private boolean doDispatch(
+        VirtualFrame frame,
+        Object self,
+        Object that,
+        Type selfType,
+        InteropConversionCallNode convertNode,
+        EqualsSimpleNode equalityNode)
+        throws PanicException {
+      var convert = UnresolvedConversion.build(selfType.getDefinitionScope());
+
+      var ctx = EnsoContext.get(this);
+      var state = State.create(ctx);
+      try {
+        var thatAsSelf = convertNode.execute(convert, state, new Object[] {selfType, that});
+        var result = equalityNode.execute(frame, self, thatAsSelf);
+        assert !result || assertHashCodeIsTheSame(self, thatAsSelf);
+        return result;
+      } catch (ArityException ex) {
+        var assertsOn = false;
+        assert assertsOn = true;
+        if (assertsOn) {
```

Comment:
Why not just panic here? `ArityException` is unexpected here, no? So it should result in some IllegalStateException

Reply:
The answer is probably the same as in [the previous discussion](https://github.com/enso-org/enso/pull/8779/files?file-filters%5B%5D=.java&show-deleted-files=true&show-viewed-files=true#r1467896387) about the topic.